src/include/sof/drivers/sai.h

```diff
@@ -0,0 +1,232 @@
+#define REG_SAI_CR2_BCP		BIT(25)
+#define REG_SAI_CR2_BCD_MSTR	BIT(24)
+#define REG_SAI_CR2_BYP		BIT(23) /* BCLK bypass */
+#define REG_SAI_CR2_DIV_MASK	0xff
+
+/* SAI Transmit and Receive Configuration 3 Register */
+#define REG_SAI_CR3_TRCE_MASK	(0xff << 16)
+#define REG_SAI_CR3_TRCE(x)	(x << 16)
+#define REG_SAI_CR3_WDFL(x)	(x)
+#define REG_SAI_CR3_WDFL_MASK	0x1f
+
+/* SAI Transmit and Receive Configuration 4 Register */
+
+#define REG_SAI_CR4_FCONT	BIT(28)
+#define REG_SAI_CR4_FCOMB_SHIFT BIT(26)
+#define REG_SAI_CR4_FCOMB_SOFT  BIT(27)
+#define REG_SAI_CR4_FCOMB_MASK  (0x3 << 26)
+#define REG_SAI_CR4_FPACK_8     (0x2 << 24)
+#define REG_SAI_CR4_FPACK_16    (0x3 << 24)
+#define REG_SAI_CR4_FRSZ(x)	(((x) - 1) << 16)
+#define REG_SAI_CR4_FRSZ_MASK	(0x1f << 16)
+#define REG_SAI_CR4_SYWD(x)	(((x) - 1) << 8)
+#define REG_SAI_CR4_SYWD_MASK	(0x1f << 8)
+#define REG_SAI_CR4_CHMOD	BIT(5)
+#define REG_SAI_CR4_CHMOD_MASK	BIT(5)
+#define REG_SAI_CR4_MF		BIT(4)
+#define REG_SAI_CR4_FSE		BIT(3)
+#define REG_SAI_CR4_FSP		BIT(1)
+#define REG_SAI_CR4_FSD_MSTR	BIT(0)
+
+/* SAI Transmit and Receive Configuration 5 Register */
+#define REG_SAI_CR5_WNW(x)	(((x) - 1) << 24)
+#define REG_SAI_CR5_WNW_MASK	(0x1f << 24)
+#define REG_SAI_CR5_W0W(x)	(((x) - 1) << 16)
+#define REG_SAI_CR5_W0W_MASK	(0x1f << 16)
+#define REG_SAI_CR5_FBT(x)	(((x) - 1) << 8)
+#define REG_SAI_CR5_FBT_MASK	(0x1f << 8)
+
+/* SAI MCLK Control Register */
+#define REG_SAI_MCTL_MCLK_EN	BIT(30)	/* MCLK Enable */
+#define REG_SAI_MCTL_MSEL_MASK	(0x3 << 24)
+#define REG_SAI_MCTL_MSEL(ID)   ((ID) << 24)
+#define REG_SAI_MCTL_MSEL_BUS	0
+#define REG_SAI_MCTL_MSEL_MCLK1	BIT(24)
+#define REG_SAI_MCTL_MSEL_MCLK2	BIT(25)
+#define REG_SAI_MCTL_MSEL_MCLK3	(BIT(24) | BIT(25))
+#define REG_SAI_MCTL_DIV_EN	BIT(23)
+#define REG_SAI_MCTL_DIV_MASK	0xFF
+
+/* SAI VERID Register */
+#define REG_SAI_VER_ID_SHIFT	16
+#define REG_SAI_VER_ID_MASK	(0xFFFF << REG_SAI_VER_ID_SHIFT)
+#define REG_SAI_VER_EFIFO_EN	BIT(0)
+#define REG_SAI_VER_TSTMP_EN	BIT(1)
+
+/* SAI PARAM Register */
+#define REG_SAI_PAR_SPF_SHIFT	16
+#define REG_SAI_PAR_SPF_MASK	(0x0F << REG_SAI_PAR_SPF_SHIFT)
+#define REG_SAI_PAR_WPF_SHIFT	8
+#define REG_SAI_PAR_WPF_MASK	(0x0F << REG_SAI_PAR_WPF_SHIFT)
+#define REG_SAI_PAR_DLN_MASK	(0x0F)
+
+/* SAI MCLK Divide Register */
+#define REG_SAI_MDIV_MASK	0xFFFFF
+
+/* SAI type */
+#define REG_SAI_DMA		BIT(0)
+#define REG_SAI_USE_AC97	BIT(1)
+#define REG_SAI_NET		BIT(2)
+#define REG_SAI_TRA_SYN		BIT(3)
+#define REG_SAI_REC_SYN		BIT(4)
+#define REG_SAI_USE_I2S_SLAVE	BIT(5)
+
+#define REG_FMT_TRANSMITTER	0
+#define REG_FMT_RECEIVER	1
+
+/* SAI clock sources */
+#define REG_SAI_CLK_BUS		0
+#define REG_SAI_CLK_MAST1	1
+#define REG_SAI_CLK_MAST2	2
+#define REG_SAI_CLK_MAST3	3
+
+#define REG_SAI_MCLK_MAX	4
+
+/* SAI data transfer numbers per DMA request */
+#define REG_SAI_MAXBURST_TX 6
+#define REG_SAI_MAXBURST_RX 6
+
+#define SAI_FLAG_PMQOS   BIT(0)
+
+#define SAI_IRQ			348
+#define EDMA_SAI_IRQ		349
```

Comment:
I have to tell you that the DMA IRQ won't work in the current form. You'll have to update src/platform/imx8/include/platform/lib/dma.h with this IRQ number properly (update my macros dma_chan_irq{,_name}). Otherwise it won't actually work.

Reply:
The values would be 29 for irq and "irqsteer5" for name. You'd have to somehow return different values depending on the channel numbers.

I think we may need to add some table or function to find this information somewhere in our platform code so it can return the correct information for both of our DAIs and the DMA channels correspoding.

Reply:
Do you confirm that moving these values in platform/imx/lib/dai,c like below would be fine ?
#define SAI1_STEER_IRQ	             348
#define EDMA_SAI1_STEER_IRQ       349
...
static struct dai sai[] = {
{
	.index = 1,
	.plat_data = {
		.base = SAI_1_BASE,
		.fifo[SOF_IPC_STREAM_PLAYBACK] = {
			.offset		= SAI_1_BASE + REG_SAI_TDR0,
			.handshake	= EDMA_HANDSHAKE( EDMA_SAI1_STEER_IRQ,
							                              EDMA_SAI_TX_CHAN),
		},
		.fifo[SOF_IPC_STREAM_CAPTURE] = {
			.offset		= SAI_1_BASE + REG_SAI_RDR0,
			.handshake	= EDMA_HANDSHAKE( EDMA_SAI1_STEER_IRQ,
							                              EDMA_SAI_RX_CHAN),
		},
	},
	.drv = &sai_driver,
},
};

Having a look tomorrow to adapt the macro dma_chan_irq{,_name}

Reply:
I'm not sure what you're doing here (it may be too late at night for me right now though). The thing is, the macros should return a valid value when the DMA domain is initialized (I think I failed to upstream this with the edma itself for some reason?) which means within platform_init().

See https://github.com/paulstelian97/sof/commit/eba4ac0825f3cc0ed6f3870085b1672d0495dc37#diff-8777c7b51425b008aa33494a7c65e59eR156 for reference on how I'm willing to initialize the platform DMA domain (that's precisely when these macros will be used, and I don't think we can defer that initialization after the platform_init, can we?)

Reply:
@groncarolo @jlacla 

When adding codes in comments please use three backticks to start the code section and three backticks to end it.
Backtick symbol is this: `

Reply:
@groncarolo We need to synchronize really. Those macros need to support both the SAI and ESAI (and, when QM comes, the second ESAI) and return different values for each.

I think a good approach would be to somehow register the interrupts (possibly add a function in the EDMA driver which is called in the probe() for both SAI and ESAI to register the interrupts) before initializing the DMA domain (I think when I will push that I can move it to the end of platform_init(), but doubt that I can do it later.

We need to discuss more on that.

Reply:
@paulstelian97 please continue with the ESAI cleanup. @groncarolo we will provide a solution for this until Monday. I might have a chance to send a patch this evening.

Reply:
fixed in #1932  and #1931